### PR TITLE
Fix up more SSE implementations for nontrapping-fp

### LINKED ISSUE #19188: Missing `wait4` stub function
When using `wait4` (or `wait3`), emscripten fails in the linking step, while perfectly finding the function header; emscripten should provide a stup-method for these like it does for `waitid` or `fork`.

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.36-git (518d9fea335f7cd2b5771e43df76e0535c6df5dd)
clang version 17.0.0 (git@github.com:llvm/llvm-project.git a7428bbb6bdb79a15e567a34d4dcc9775454ed71)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /path/to/llvm-project/build/bin
```

**Failing command line in full:**
```
$ /path/to/emscripten/em++ -o wait4.js wait4.cpp
wasm-ld: error: /tmp/emscripten_temp_3a84yme1/wait4_0.o: undefined symbol: wait4
em++: error: '/path/to/llvm-project/build/bin/wasm-ld -o wait4.wasm /tmp/emscripten_temp_3a84yme1/wait4_0.o -L/path/to/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-debug-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/tmp/tmpwfq8ycsv.undefined --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=__get_temp_ret --export=__set_temp_ret --export=__wasm_call_ctors --export-table -z stack-size=65536 --initial-memory=16777216 --no-entry --max-memory=16777216 --stack-first' failed (returned 1)
```

**Full link command and output with `-v` appended:**
```
$ /path/to/emscripten/em++ -v -o wait4.js wait4.cpp
 "/path/to/llvm-project/build/bin/clang" --version
 "/path/to/llvm-project/build/bin/clang++" -target wasm32-unknown-emscripten -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -DEMSCRIPTEN --sysroot=/path/to/emscripten/cache/sysroot -Xclang -iwithsysroot/include/fakesdl -Xclang -iwithsysroot/include/compat -v wait4.cpp -c -o /tmp/emscripten_temp_3xlnu5mg/wait4_0.o
clang version 17.0.0 (git@github.com:llvm/llvm-project.git a7428bbb6bdb79a15e567a34d4dcc9775454ed71)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /path/to/llvm-project/build/bin
 (in-process)
 "/path/to/llvm-project/build/bin/clang-17" -cc1 -triple wasm32-unknown-emscripten -emit-obj -mrelax-all -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name wait4.cpp -mrelocation-model static -mframe-pointer=none -ffp-contract=on -fno-rounding-math -mconstructor-aliases -target-cpu generic -debugger-tuning=gdb -v -fcoverage-compilation-dir=/path/to/test/project -resource-dir /path/to/llvm-project/build/lib/clang/17 -D EMSCRIPTEN -isysroot /path/to/emscripten/cache/sysroot -internal-isystem /path/to/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1 -internal-isystem /path/to/emscripten/cache/sysroot/include/c++/v1 -internal-isystem /path/to/llvm-project/build/lib/clang/17/include -internal-isystem /path/to/emscripten/cache/sysroot/include/wasm32-emscripten -internal-isystem /path/to/emscripten/cache/sysroot/include -fdeprecated-macro -fdebug-compilation-dir=/path/to/test/project -ferror-limit 19 -fvisibility=default -fgnuc-version=4.2.1 -fcxx-exceptions -fignore-exceptions -fexceptions -fcolor-diagnostics -iwithsysroot/include/fakesdl -iwithsysroot/include/compat -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -o /tmp/emscripten_temp_3xlnu5mg/wait4_0.o -x c++ wait4.cpp
clang -cc1 version 17.0.0 based upon LLVM 17.0.0git default target x86_64-unknown-linux-gnu
ignoring nonexistent directory "/path/to/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1"
ignoring nonexistent directory "/path/to/emscripten/cache/sysroot/include/wasm32-emscripten"
#include "..." search starts here:
#include <...> search starts here:
 /path/to/emscripten/cache/sysroot/include/fakesdl
 /path/to/emscripten/cache/sysroot/include/compat
 /path/to/emscripten/cache/sysroot/include/c++/v1
 /path/to/llvm-project/build/lib/clang/17/include
 /path/to/emscripten/cache/sysroot/include
End of search list.
 "/path/to/llvm-project/build/bin/wasm-ld" -o wait4.wasm /tmp/emscripten_temp_3xlnu5mg/wait4_0.o -L/path/to/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-debug-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/tmp/tmpef8yn14z.undefined --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=__get_temp_ret --export=__set_temp_ret --export=__wasm_call_ctors --export-table -z stack-size=65536 --initial-memory=16777216 --no-entry --max-memory=16777216 --stack-first
wasm-ld: error: /tmp/emscripten_temp_3xlnu5mg/wait4_0.o: undefined symbol: wait4
em++: error: '/path/to/llvm-project/build/bin/wasm-ld -o wait4.wasm /tmp/emscripten_temp_3xlnu5mg/wait4_0.o -L/path/to/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-debug-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --allow-undefined-file=/tmp/tmpef8yn14z.undefined --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_get_current --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=__get_temp_ret --export=__set_temp_ret --export=__wasm_call_ctors --export-table -z stack-size=65536 --initial-memory=16777216 --no-entry --max-memory=16777216 --stack-first' failed (returned 1)
```

### LINKED ISSUE #20279: Combining wasm workers and pthreads leads to excessive worker creation
I am using both pthreads and wasm workers, and I see more workers being created than necessary.
The reason seems to be the following:

`PThread.init` checks for `ENVIRONMENT_IS_PTHREAD` first, in all other cases `Pthread.initMainThread` is called. To me it looks like there should be an additional check for `ENVIRONMENT_IS_WASM_WORKER` in the generated js:

```
 init: function() {
  PThread.debugInit();
  if (ENVIRONMENT_IS_PTHREAD || ENVIRONMENT_IS_WASM_WORKER) {
   PThread.initWorker();
  } else {
   PThread.initMainThread();
  }
 },

```
The current init function leads to `initMainThread` being called for each wasm worker, which in turn will call `PThread.init` to allocate the number of workers specified by `PTHREAD_POOL_SIZE` multiple times..
It seems to work correctly with the suggested change, however, I am not familiar enough with the internals of Emscripten to make a judgement.

**Version of emscripten/emsdk:** 3.1.46

**Command line in full:**
emcc ./src/main.c ...list if source files
-I ./include/ ^
-g3 ^
--source-map-base ./ ^
-gsource-map ^
-s ALLOW_MEMORY_GROWTH=1 ^
-s ENVIRONMENT=web,worker ^
--shell-file ./index_template.html ^
-s SUPPORT_ERRNO=0 ^
-s MODULARIZE=1 ^
-sMALLOC=emmalloc ^
-sWASM_WORKERS ^
-sOFFSCREEN_FRAMEBUFFER ^
-pthread ^
-s PTHREAD_POOL_SIZE=2 ^
-s "EXPORT_NAME='wasmMod'" ^
-s EXPORTED_FUNCTIONS="['_malloc','_free','_main']" ^
-s EXPORTED_RUNTIME_METHODS="['cwrap','UTF16ToString','UTF8ToString','stringToUTF8','allocateUTF8','AsciiToString']" ^
-o index.html

### PR DESCRIPTION
**Pull request by dschuff**
Fixes lto2.test_sse1 and test_sse2 with checks similar to #22911 and #22893

### CODE REVIEW

File: system/include/compat/xmmintrin.h [hunk 1]
```diff
@@ -596,8 +596,9 @@ _mm_cvtsi32_ss(__m128 __a, int __b)
 
 static __inline__ int __attribute__((__always_inline__, __nodebug__, DIAGNOSE_SLOW)) _mm_cvtss_si32(__m128 __a)
 {
-  int x = lrint(((__f32x4)__a)[0]);
-  if (x != 0 || fabsf(((__f32x4)__a)[0]) < 2.f)
+  float e = ((__f32x4)__a)[0];
+  int x = lrint(e);
+  if ((x != 0 || fabsf(e)) < 2.f && !isnan(e) && e <= (float)INT_MAX && e >= INT_MIN)
```
Comment (sbc100) [inline system/include/compat/xmmintrin.h:601]: Can these comparisons be made against `x` to avoid the cast here?

Is the cast needed?

Comment (dschuff) [reply system/include/compat/xmmintrin.h:601]: without the explicit cast, there's a warning about the implicit cast changing the value of INT_MAX.
Although looking at the other implementations in this file, they also have similar implicit casts which also produce that warning, so maybe we should be consistent.

Comment (dschuff) [reply system/include/compat/xmmintrin.h:601]: And we can't compare against x (at least like it's written here) because an int is always <= INT_MAX

Comment (dschuff) [reply system/include/compat/xmmintrin.h]: I found that the comparison style (e.g. whether the nan check was first or further down) was inconsistent so I made all the checks the same. I think if we want to optimize this further we should look more into using the nontrapping instructions directly but it's not clear whether that's important.

**Review by sbc100 — APPROVED** [review]
Can you add the test you are fixing to `test-core3` in circleci config?

**Pull request merged**